examples/wmt/train.py

```diff
@@ -0,0 +1,638 @@
+      metrics_all = common_utils.get_metrics(metrics_all)
+      lr = metrics_all.pop('learning_rate').mean()
+      metrics_sums = jax.tree_map(jnp.sum, metrics_all)
+      denominator = metrics_sums.pop('denominator')
+      summary = jax.tree_map(lambda x: x / denominator, metrics_sums)  # pylint: disable=cell-var-from-loop
+      summary['learning_rate'] = lr
+      logging.info('train in step: %d, loss: %.4f', step, summary['loss'])
+      steps_per_eval = FLAGS.eval_frequency if step != 0 else 1
+      steps_per_sec = steps_per_eval / (time.time() - t_loop_start)
+      t_loop_start = time.time()
+      if jax.host_id() == 0:
+        train_summary_writer.scalar('steps per second', steps_per_sec, step)
+        for key, val in summary.items():
+          train_summary_writer.scalar(key, val, step)
+        train_summary_writer.flush()
+      metrics_all = []
+
+      # Eval Metrics
+      t_eval_start = time.time()
+      eval_metrics = []
+      eval_iter = iter(eval_ds)
+      for _, eval_batch in zip(range(FLAGS.num_eval_steps), eval_iter):
+        eval_batch = jax.tree_map(lambda x: x._numpy(), eval_batch)  # pylint: disable=protected-access
+        eval_batch = common_utils.shard(eval_batch)
+        metrics = p_eval_step(optimizer.target, eval_batch)
+        eval_metrics.append(metrics)
+      eval_metrics = common_utils.get_metrics(eval_metrics)
+      eval_metrics_sums = jax.tree_map(jnp.sum, eval_metrics)
+      eval_denominator = eval_metrics_sums.pop('denominator')
+      eval_summary = jax.tree_map(
+          lambda x: x / eval_denominator,  # pylint: disable=cell-var-from-loop
+          eval_metrics_sums)
+      logging.info('eval in step: %d, loss: %.4f', step, eval_summary['loss'])
+      if jax.host_id() == 0:
+        for key, val in eval_summary.items():
+          eval_summary_writer.scalar(key, val, step)
+        eval_summary_writer.flush()
+      logging.info('eval time: %.4f s step %d', time.time()-t_eval_start, step)
+
+      # Translation and BLEU Score.
+      t_inference_start = time.time()
+      predict_iter = iter(predict_ds)
+      sources, references, predictions = [], [], []
+      for _, pred_batch in enumerate(predict_iter):
+        pred_batch = jax.tree_map(lambda x: x._numpy(), pred_batch)  # pylint: disable=protected-access
+        # Handle final odd-sized batch by padding instead of dropping it.
+        cur_pred_batch_size = pred_batch['inputs'].shape[0]
+        if cur_pred_batch_size % n_devices:
+          logging.info('Translation: uneven batch size %d.',
+                       cur_pred_batch_size)
+          padded_size = int(
+              np.ceil(cur_pred_batch_size / n_devices) * n_devices)
+          pred_batch = jax.tree_map(
+              lambda x: pad_examples(x, padded_size), pred_batch)  # pylint: disable=cell-var-from-loop
+        pred_batch = common_utils.shard(pred_batch)
+        per_device_batchsize = pred_batch['inputs'].shape[1]
+        cache = jax_utils.replicate(
```

Comment:
What is the intuition behind "cache"? Why do we only need it for prediction?

Reply:
The cache is a piece of state that builds up the attention "keys" and "values" one position at a time during a fast-decoding loop - it's a crucial performance optimization needed for doing inference on an autoregressive model.  Otherwise a naive sampling loop tries to constantly recalculate these same values over and over for each new predicted position.